Galaxus.Functional/(Option)/(Extensions)/OptionLinqExtensions.cs

```diff
@@ -0,0 +1,143 @@
+        ///     <see cref="IEnumerable{T}" />.
+        /// </returns>
+        /// <exception cref="ArgumentNullException"><c>source</c> is <c>null</c>.</exception>
+        public static Option<TSource> LastOrNone<TSource>(this IEnumerable<TSource> source)
+        {
+            var value = source.ToList();
+            return value.Count == 0 ? Option<TSource>.None : Option<TSource>.Some(value[value.Count - 1]);
+        }
+
+        /// <summary>
+        ///     Returns the last element of a sequence that satisfies a condition or <see cref="Option{T}.None" /> if no such
+        ///     element is found.
+        /// </summary>
+        /// <typeparam name="TSource">The type of the elements of <c>source</c>.</typeparam>
+        /// <param name="source">An <see cref="IEnumerable{T}" /> to return an element from.</param>
+        /// <param name="predicate">A function to test each element for a condition.</param>
+        /// <returns>
+        ///     <c><see cref="Option{T}" />.None</c> if the sequence is empty or if no elements pass the test in the predicate
+        ///     function; otherwise, the last element that passes the test in the predicate function.
+        /// </returns>
+        /// <exception cref="ArgumentNullException"><c>source</c> or <c>predicate</c> is <c>null</c>.</exception>
+        public static Option<TSource> LastOrNone<TSource>(this IEnumerable<TSource> source,
+            Func<TSource, bool> predicate)
+        {
+            var value = source.Where(predicate).ToList();
+            return value.Count == 0 ? Option<TSource>.None : Option<TSource>.Some(value[value.Count - 1]);
+            //var value = source.LastOrDefault(predicate);
+            //return value == null ? Option<TSource>.None : Option<TSource>.Some(value);
+        }
+
+        /// <summary>
+        ///     Returns the only element of a sequence, or <see cref="Option{T}.None" /> if the sequence is empty; this method
+        ///     throws an exception if there is more than one element in the sequence.
+        /// </summary>
+        /// <typeparam name="TSource">The type of the elements of <c>source</c>.</typeparam>
+        /// <param name="source">An <see cref="IEnumerable{T}" /> to return the single element of.</param>
+        /// <returns>
+        ///     The single element of the input sequence, or
+        ///     <c>
+        ///         <see cref="Option{T}.None" />
+        ///     </c>
+        ///     if the sequence contains no elements.
+        /// </returns>
+        /// <exception cref="ArgumentNullException"><c>source</c> is <c>null</c>.</exception>
+        /// <exception cref="InvalidOperationException">The input sequence contains more than one element.</exception>
+        public static Option<TSource> SingleOrNone<TSource>(this IEnumerable<TSource> source)
+        {
+            var value = source.ToList();
+            return value.Count == 0 ? Option<TSource>.None :
```

Comment:
Oh, a nested ternary. Didn't even see that - I always recommend putting ternaries on a single line (again this kind of comment!) or formatting them like

```csharp
IsFoo()
    ? "foo"
    : "bar"
```

But I feel like the "more than 1 match"-case should not be a part of the return... it's a bit different than the rest, throwing an exception and stuff. Would break up the nested ternary as well, so it's win-win! :D